File: grim/src/switch.rs
use std::env::consts::OS;
use std::fs;

use crate::util::{self, error, exec};

const OUT_LINK: &str = "/tmp/nix-build-result";

pub fn switch(host: &Option<String>) {
    let platform = match OS {
        "macos" => "darwin",
        _ => "nixos",
    };

    let host = match host {
        Some(val) => val,
        None => "$(hostname)",
    };

    nix_build(platform, host);
    nix_switch(platform, host);

    match fs::remove_file(OUT_LINK) {
        Ok(_) => (),
        Err(err) => error(err),
    }
}

fn nix_build(platform: &str, host: &str) {
    let cmd = format!(
        "nix build {}/#{}Configurations.{}.system --out-link {}",
        system_dir(),
        platform,
        host,
        OUT_LINK
    );
    exec(cmd);
}

fn nix_switch(platform: &str, host: &str) {
    let cmd = match platform {
        "darwin" => darwin_switch_cmd(host),
        "nixos" => nixos_switch_cmd(host),
        _ => String::from("echo 'Unsupported operating system!"),
    };
    exec(cmd);
}

fn darwin_switch_cmd(host: &str) -> String {
    format!(
        "{}/sw/bin/darwin-rebuild switch --flake {}#{}",
        OUT_LINK,
        system_dir(),
        host
    )
}

fn nixos_switch_cmd(host: &str) -> String {
    format!(
        "sudo nixos-rebuild switch --flake {}#{}",
        system_dir(),
        host
    )
}

fn system_dir() -> String {
    String::from(util::system_dir().to_str().unwrap())
}

File: grim/cmd/switch.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"runtime"

	"github.com/spf13/cobra"
)

var host string
var outdir string = fmt.Sprintf("%snix-build-result", os.TempDir())

var switchCmd = &cobra.Command{
	Use:   "switch",
	Short: "Switch to a new Nix system generation",
	Run: func(cmd *cobra.Command, args []string) {
		platform := runtime.GOOS
		nix_switch(platform, host)
	},
}

func nix_switch(platform string, host string) {
	var cmd *exec.Cmd
	if platform == "darwin" {
		cmd = darwin_switch_cmd(host)
	} else {
		cmd = nixos_switch_cmd(host)
	}
	run_cmd(cmd)
}

func run_cmd(cmd *exec.Cmd) {
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err := cmd.Run()
	if err != nil {
		fmt.Println(err)
	}
}

func darwin_switch_cmd(host string) *exec.Cmd {
	flake := fmt.Sprintf("%s#%s", SystemDir(), host)
	return exec.Command("darwin-rebuild", "switch", "--flake", flake)
}

func nixos_switch_cmd(host string) *exec.Cmd {
	flake := fmt.Sprintf("%s#%s", SystemDir(), host)
	return exec.Command("sudo", "nixos-rebuild", "switch", "--flake", flake)
}

func init() {
	hostname, err := os.Hostname()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	switchCmd.Flags().StringVarP(&host, "host", "H", hostname, "Nix host")
	rootCmd.AddCommand(switchCmd)
}
